File: frontend/src/app/features/electrical/ohms-law/ohms-law.ts
import { Component } from '@angular/core';
import {
  FormBuilder,
  ReactiveFormsModule,
} from '@angular/forms';

import { CalculationService } from '../../../services/calculation.service';
import {
  CalculationRequest,
  CalculationResponse,
} from '../../../core/models/calculation.model';


@Component({
  selector: 'app-ohms-law',
  imports: [ReactiveFormsModule],
  templateUrl: './ohms-law.html',
  styleUrl: './ohms-law.css',
})
export class OhmsLaw {

  calculationResult: CalculationResponse | null = null;

  calculationError: string | null = null;

  isCalculating = false;


  readonly calculatorForm;


  constructor(
    private readonly formBuilder: FormBuilder,
    private readonly calculationService: CalculationService,
  ) {

    this.calculatorForm = this.formBuilder.group({
      operation: ['calculate-current'],

      voltage: [null],
      current: [null],
      resistance: [null],
    });
  }


  get selectedOperation(): string {
    return this.calculatorForm.value.operation ?? '';
  }


  calculate(): void {

    this.calculationResult = null;
    this.calculationError = null;

    const formValue = this.calculatorForm.value;

    const values: Record<string, number> = {};

    if (this.selectedOperation === 'calculate-voltage') {
      values['current'] = Number(formValue.current);
      values['resistance'] = Number(formValue.resistance);
    }

    if (this.selectedOperation === 'calculate-current') {
      values['voltage'] = Number(formValue.voltage);
      values['resistance'] = Number(formValue.resistance);
    }

    if (this.selectedOperation === 'calculate-resistance') {
      values['voltage'] = Number(formValue.voltage);
      values['current'] = Number(formValue.current);
    }


    const request: CalculationRequest = {
      operation: this.selectedOperation,
      values,
    };


    this.isCalculating = true;


    this.calculationService
      .calculateOhmsLaw(request)
      .subscribe({

        next: (response) => {
          this.calculationResult = response;
          this.isCalculating = false;
        },

        error: (error) => {
          this.calculationError =
            error?.error?.detail ??
            'Unable to calculate the result.';

          this.isCalculating = false;
        },

      });
  }
}

File: backend/app/calculators/electrical/ohms_law.py
def calculate_ohms_law(operation: str, values: dict[str, float]):
    if operation == "calculate-voltage":
        current = values.get("current")
        resistance = values.get("resistance")

        if current is None or resistance is None:
            raise ValueError(
                "Current and resistance are required to calculate voltage."
            )

        result = current * resistance

        return {
            "formula": "V = I × R",
            "result": result,
            "unit": "V",
            "steps": [
                "V = I × R",
                f"V = {current} × {resistance}",
                f"V = {result} V",
            ],
        }

    elif operation == "calculate-current":
        voltage = values.get("voltage")
        resistance = values.get("resistance")

        if voltage is None or resistance is None:
            raise ValueError(
                "Voltage and resistance are required to calculate current."
            )

        if resistance == 0:
            raise ValueError(
                "Resistance cannot be zero."
            )

        result = voltage / resistance

        return {
            "formula": "I = V / R",
            "result": result,
            "unit": "A",
            "steps": [
                "I = V / R",
                f"I = {voltage} / {resistance}",
                f"I = {result} A",
            ],
        }

    elif operation == "calculate-resistance":
        voltage = values.get("voltage")
        current = values.get("current")

        if voltage is None or current is None:
            raise ValueError(
                "Voltage and current are required to calculate resistance."
            )

        if current == 0:
            raise ValueError(
                "Current cannot be zero."
            )

        result = voltage / current

        return {
            "formula": "R = V / I",
            "result": result,
            "unit": "Ω",
            "steps": [
                "R = V / I",
                f"R = {voltage} / {current}",
                f"R = {result} Ω",
            ],
        }

    else:
        raise ValueError(
            "Invalid Ohm's Law operation."
        )